homeassistant/components/imap/coordinator.py

```diff
@@ -34,75 +36,131 @@ async def connect_to_server(data: Mapping[str, Any]) -> IMAP4_SSL:
-            )
         return len(lines[0].split())
 
-    async def async_wait_server_push(self) -> None:
-        """Wait for data push from server."""
-        try:
-            idle: asyncio.Future = await self.imap_client.idle_start()
-            await self.imap_client.wait_server_push()
-            self.imap_client.idle_done()
-            async with async_timeout.timeout(10):
-                await idle
-
-        except (AioImapException, asyncio.TimeoutError):
-            _LOGGER.warning(
-                "Lost %s (will attempt to reconnect)",
-                self.config_entry.data[CONF_SERVER],
-            )
+    async def _cleanup(self) -> None:
+        """Close resources."""
+        if self.imap_client:
+            try:
+                if self.imap_client.has_pending_idle():
+                    self.imap_client.idle_done()
+                await self.imap_client.stop_wait_server_push()
+                await self.imap_client.logout()
+                await self.imap_client.close()
+            except AioImapException:
+                _LOGGER.warning("Error while cleaning up imap connection")
             self.imap_client = None
-        await self.async_request_refresh()
 
     async def shutdown(self, *_) -> None:
         """Close resources."""
-        if self.imap_client:
-            if self.imap_client.has_pending_idle():
+        await self._cleanup()
+
+
+class ImapPollingDataUpdateCoordinator(ImapDataUpdateCoordinator):
+    """Class for imap client."""
+
+    def __init__(self, hass: HomeAssistant, imap_client: IMAP4_SSL) -> None:
+        """Initiate imap client."""
+        super().__init__(hass, imap_client, timedelta(seconds=10))
+
+    async def _async_update_data(self) -> int | None:
+        """Update the number of unread emails."""
+        return await self._async_fetch_number_of_messages()
+
+
+class ImapPushDataUpdateCoordinator(ImapDataUpdateCoordinator):
+    """Class for imap client."""
+
+    def __init__(self, hass: HomeAssistant, imap_client: IMAP4_SSL) -> None:
+        """Initiate imap client."""
+        super().__init__(hass, imap_client, None)
+        self._push_wait_task: asyncio.Task[None] | None = None
+
+    async def _async_update_data(self) -> int | None:
+        """Update the number of unread emails."""
+        await self.async_start()
+        return None
+
+    async def async_start(self) -> None:
+        """Start coordinator."""
+        self._push_wait_task = self.hass.async_create_background_task(
+            self._async_wait_push_loop(), "Wait for IMAP data push"
+        )
+
+    async def _async_wait_push_loop(self) -> None:
+        """Wait for data push from server."""
+        while True:
+            await self._async_reconnect_if_needed()
+            try:
+                number_of_messages = await self._async_fetch_number_of_messages()
+            except UpdateFailed as ex:
+                self.async_set_update_error(ex)
```

Comment:
Should we do a cleanup here in case the failure is because the connection is in a bad state?

Reply:
```
                await self._cleanup()
                await asyncio.sleep(BACKOFF_TIME)
		continue
```